import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java

```diff
@@ -82,70 +93,107 @@ protected void store(
-      throw serviceException;
+        if(destinationMap != null){
+          DicomInputStream inDicomStream  = new DicomInputStream(inPdvStream);
+          inDicomStream.mark(Integer.MAX_VALUE);
+          Attributes attrs = inDicomStream.readDataset(-1, Tag.PixelData);
+          inDicomStream.reset();
+
+          countingStream = new CountingInputStream(inDicomStream);
+          destinationClient.set(selectDestinationClient(association.getAAssociateAC().getCallingAET(), attrs));
+        } else {
+          countingStream = new CountingInputStream(inPdvStream);
+          destinationClient.set(defaultDicomWebClient);
+        }
+
+        InputStream inWithHeader =
+            DicomStreamUtil.dicomStreamWithFileMetaHeader(
+                sopInstanceUID, sopClassUID, transferSyntax, countingStream);
+
+        List<StreamProcessor> processorList = new ArrayList<>();
+        if (redactor != null) {
+          processorList.add(redactor::redact);
+        }
+
+        if(transcodeToSyntax != null && !transcodeToSyntax.equals(transferSyntax)) {
+          processorList.add((inputStream, outputStream) -> {
+            try (Transcoder transcoder = new Transcoder(inputStream)) {
+              transcoder.setIncludeFileMetaInformation(true);
+              transcoder.setDestinationTransferSyntax(transcodeToSyntax);
+              transcoder.transcode((transcoder1, dataset) -> outputStream);
+            }
+          });
+        }
+
+        if (backupUploadService != null) {
+          processorList.add((inputStream, outputStream) -> {
+            byte[] bytes;
+            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+              StreamUtils.copy(inputStream, baos);
+              bytes = baos.toByteArray();
+              backupState.set(backupUploadService.createBackup(bytes, sopInstanceUID));
+            } catch (IOException ioex) {
+              log.error("Backup creation failed.", ioex);
+              throw new IBackupUploader.BackupException("Backup creation failed.", ioex);
+            }
+
+            try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes)) {
+              StreamUtils.copy(bais, outputStream);
+            }
+          });
+        }
+
+        processorList.add((inputStream, outputStream) -> {
+          destinationClient.get().stowRs(inputStream);
+        });
+
+        processStream(association.getApplicationEntity().getDevice().getExecutor(),
+            inWithHeader, processorList);
+
+        response.setInt(Tag.Status, VR.US, Status.Success);
+        MonitoringService.addEvent(Event.CSTORE_BYTES, countingStream.getCount());
+      } catch (DicomWebException e) {
+      if (backupUploadService != null) {
+        MonitoringService.addEvent(Event.CSTORE_BACKUP_ERROR);
+        log.error("C-STORE request failed. Trying to resend...", e);
+        firstUploadedAttemptFailed = true;
+        backupUploadService.startUploading(destinationClient.get(), backupState.get());
```

Comment:
If the backup service eventually succeeds, we should mark the response as successful: response.setInt(Tag.Status, VR.US, Status.Success);